apps/bfd-pipeline/bfd-pipeline-shared-utils/src/main/java/gov/cms/bfd/pipeline/sharedutils/samhsa/backfill/AbstractSamhsaBackfill.java

```diff
@@ -0,0 +1,278 @@
+          + " VALUES (:tableName, :lastClaim) "
+          + " ON CONFLICT (claim_table) "
+          + " DO UPDATE SET "
+          + "   last_processed_claim = :lastClaim ";
+
+  /** Query to get the backfill progress from the database for a given claim table. */
+  protected String GET_PROGRESS_QUERY =
+      " SELECT last_processed_claim FROM ccw.samhsa_backfill_progress "
+          + " WHERE claim_table = :tableName ";
+
+  /** The transaction manager. */
+  TransactionManager transactionManager;
+
+  /**
+   * SamhsaUtil class. This will be used to check the claims for SAMHSA data, and create the tags if
+   * necessary.
+   */
+  SamhsaUtil samhsaUtil;
+
+  /** query batch size. */
+  int batchSize;
+
+  /**
+   * Constructor.
+   *
+   * @param transactionManager The transaction manager.
+   * @param batchSize the query batch size. This is the limit of claims to be pulled with each
+   *     query.
+   * @param logger The logger.
+   * @param tableEntry the table Entry for this thread.
+   */
+  public AbstractSamhsaBackfill(
+      TransactionManager transactionManager, int batchSize, Logger logger, TableEntry tableEntry) {
+    this.logger = logger;
+    this.transactionManager = transactionManager;
+    this.batchSize = batchSize;
+    this.tableEntry = tableEntry;
+    samhsaUtil = SamhsaUtil.getSamhsaUtil();
+  }
+
+  /**
+   * Executes the query using the transaction manager.
+   *
+   * @param query The query to execute.
+   * @return a list of claims.
+   * @param <TClaim> The type of the claim.
+   */
+  private <TClaim> List<TClaim> executeQuery(Query query) {
+    return (List<TClaim>) query.getResultList();
+  }
+
+  /**
+   * Builds a query object by taking a SQL query string and replacing the parameter placeholders
+   * with appropriate values.
+   *
+   * @param startingClaim The claim to start at. If this is empty, the version of the query with no
+   *     starting claim will be used.
+   * @param tableEntry The TableEntry pojo for this table.
+   * @param limit the number of claims to pull at a time.
+   * @param entityManager The entity manager.
+   * @return Query to run.
+   */
+  private Query buildQuery(
+      Optional<String> startingClaim,
+      TableEntry tableEntry,
+      int limit,
+      EntityManager entityManager) {
+    Map<String, String> params =
+        Map.of(
+            "tableName",
+            tableEntry.getClaimTable(),
+            "claimColumn",
+            tableEntry.getClaimColumnName(),
+            "startingClaim",
+            startingClaim.orElse(""),
+            "tagTableName",
+            tableEntry.getTagTable(),
+            "limit",
+            String.valueOf(limit));
+    StringSubstitutor strSub = new StringSubstitutor(params);
+    String queryStr =
+        strSub.replace(
+            startingClaim.isPresent() ? QUERY_WITH_STARTING_CLAIM : QUERY_WITH_NO_STARTING_CLAIM);
+    return entityManager.createNativeQuery(queryStr, tableEntry.getClaimClass());
```

Comment:
Would it be possible to use `createQuery` instead of `createNativeQuery`? That will at least perform some type checking to ensure the type is assignable to the query result. I believe that would remove the need for `select *` as well.

Reply:
Yeah, we can do that. The check for existing tags may be a bit tricky, but I'll see what I can do.

Reply:
I think these changes should work fine. Testing performance now.